File: BlotoutFoundation/BlotoutFoundationAll/BOFUtilities/BOFUserDefaults.m
//
//  BOFUserDefaults.m
//  BlotoutFoundation
//
//  Created by Blotout on 28/07/19.
//  Copyright © 2019 Blotout. All rights reserved.
//

#import "BOFUserDefaults.h"
#import "BOFLogs.h"
#include "BOFConstants.h"
#import "BlotoutFoundation.h"

@interface BOFUserDefaults()
@property(nonatomic,strong)NSString *productKey;
@property(atomic,strong)NSMutableDictionary *productContainer;
@end


static NSMutableDictionary *sBOFDefaultMap = nil;
static dispatch_queue_t sBOFDefaultsSerialQueue = nil;

@implementation BOFUserDefaults

+(void)initialize{
    @try {
        sBOFDefaultMap = [NSMutableDictionary dictionary];
        sBOFDefaultsSerialQueue = dispatch_queue_create(BO_SDK_DEFAULT_QUEUE, DISPATCH_QUEUE_SERIAL);
        
    } @catch (NSException *exception) {
         BOFLogDebug(@"%@:%@", BOF_DEBUG, exception);
    }
}

-(instancetype)initWithProduct:(NSString *)key{
    @try {
        self = [super init];
        if ( self ) {
            self.productKey = key;
        }
        return self;
    } @catch (NSException *exception) {
         BOFLogDebug(@"%@:%@", BOF_DEBUG, exception);
    }
    return nil;
}

+(_Nonnull instancetype)userDefaultsForProduct:( NSString * _Nonnull )product{
    @try {
        __block BOFUserDefaults *defaultsInstance = nil;
        dispatch_sync(sBOFDefaultsSerialQueue, ^{
            defaultsInstance = (BOFUserDefaults *)[sBOFDefaultMap objectForKey:product];
            if ( defaultsInstance == nil ) {
                defaultsInstance = [[BOFUserDefaults alloc] initWithProduct:product];
                [sBOFDefaultMap setObject:defaultsInstance forKey:product];
            }
        });
        return defaultsInstance;
    } @catch (NSException *exception) {
         BOFLogDebug(@"%@:%@", BOF_DEBUG, exception);
    }
    return nil;
}

+(void)root:(void( ^ _Nonnull )(NSMutableDictionary * _Nonnull root ))updateBlock
{
    @try {
        NSUserDefaults *ud = [NSUserDefaults standardUserDefaults];
        
        NSDictionary *rootImmutable = [ud dictionaryForKey:BO_SDK_ROOT_USER_DEFAULTS_KEY];
        
        if ( rootImmutable == nil  ) {
            rootImmutable = @{};
        }
        
        NSMutableDictionary *rootMutable = [rootImmutable mutableCopy];
        
        updateBlock(rootMutable);
        
        [ud setObject:rootMutable forKey:BO_SDK_ROOT_USER_DEFAULTS_KEY];
        
        [ud synchronize];
    } @catch (NSException *exception) {
         BOFLogDebug(@"%@:%@", BOF_DEBUG, exception);
    }
}


-(void)updateDefaultForProduct:( BOOL (^ _Nonnull )(NSMutableDictionary * _Nonnull produtContainer) )updateBlock{
    @try {
        __weak BOFUserDefaults *weakSelf = self;
        [BOFUserDefaults root:^(NSMutableDictionary * _Nonnull root) {
            
            //Get the product level defaults
            //Eg:
            // com.blotout.root.sdk
            //      -- Product1
            //          --Product1 defaults
            //      -- Product2
            //          --Product2 defaults
            NSMutableDictionary *productContainer = [[root objectForKey:weakSelf.productKey] mutableCopy];
            if ( productContainer == nil ) {
                productContainer = [@{} mutableCopy];
            }
            
            //have the defaults write to product container
            BOOL hasChanged = updateBlock(productContainer);
            if (hasChanged ) {
                [root setObject:productContainer forKey:weakSelf.productKey];
            }
            
        }];
    } @catch (NSException *exception) {
         BOFLogDebug(@"%@:%@", BOF_DEBUG, exception);
    }
}


-(void)setObject:(nullable id)obj forKey:(nonnull id<NSCopying>)aKey{
    @try {
        if ( obj == nil ) {
            return;
        }
        
        //We have batch updates in progress
        if ( self.productContainer != nil ) {
            [self.productContainer setObject:obj forKey:aKey];
        }
        else{
            [self updateDefaultForProduct:^BOOL(NSMutableDictionary * _Nonnull produtContainer) {
                [produtContainer setObject:obj forKey:aKey];
                return TRUE;
            }];
        }
    } @catch (NSException *exception) {
         BOFLogDebug(@"%@:%@", BOF_DEBUG, exception);
    }
}

-(void)removeObjectForKey:(nonnull id<NSCopying>)aKey{
    @try {
        if ( self.productContainer != nil ) {
            [self.productContainer removeObjectForKey:aKey];
        }
        else{
            [self updateDefaultForProduct:^BOOL(NSMutableDictionary * _Nonnull produtContainer) {
                [produtContainer removeObjectForKey:aKey];
                return TRUE;
            }];
        }
    } @catch (NSException *exception) {
         BOFLogDebug(@"%@:%@", BOF_DEBUG, exception);
    }

}

-(nullable id)objectForKey:(nonnull id<NSCopying>)key{
    @try {
        __block id object = nil;
        
        if (self.productContainer != nil ) {
            object = [self.productContainer objectForKey:key];
        }
        else{
            [self updateDefaultForProduct:^BOOL(NSMutableDictionary * _Nonnull produtContainer) {
                object = [produtContainer objectForKey:key];
                return NO;
            }];
        }
        
        return object;
    } @catch (NSException *exception) {
         BOFLogDebug(@"%@:%@", BOF_DEBUG, exception);
    }
    return nil;
}

-(void)batchUpdates:( void(^ _Nonnull )(BOFUserDefaults *  _Nonnull defaults))updateBlock{
    @try {
        [self updateDefaultForProduct:^BOOL(NSMutableDictionary * _Nonnull produtContainer) {
            self.productContainer = produtContainer;
            updateBlock(self);
            self.productContainer = nil;
            return TRUE;
        }];
    } @catch (NSException *exception) {
         BOFLogDebug(@"%@:%@", BOF_DEBUG, exception);
    }
}

@end

File: BlotoutAnalytics/Foundation/BOFUtilities/BOFUserDefaults.swift
//
//  BOFUserDefaults.swift
//  BlotoutAnalyticsSDK
//
//  Created by Poonam Tiwari on 31/03/22.
//

import Foundation
private var sBOFDefaultMap: [AnyHashable : Any]? = nil
private var sBOFDefaultsSerialQueue: DispatchQueue? = nil

class BOFUserDefaults:NSObject {
    private var productKey: String?
    private var productContainer: [AnyHashable : Any]?
    
    
    init(product key: String?) {
        super.init()
        productKey = key
        sBOFDefaultMap = [AnyHashable : Any]()
        sBOFDefaultsSerialQueue = DispatchQueue(label: BO_SDK_DEFAULT_QUEUE)
    }
    
    class func userDefaults(forProduct product: String) -> Self {
        var defaultsInstance: BOFUserDefaults? = nil
        sBOFDefaultsSerialQueue?.sync(execute: {
            defaultsInstance = sBOFDefaultMap?[product] as? BOFUserDefaults
            if defaultsInstance == nil {
                defaultsInstance = BOFUserDefaults(product: product)
                sBOFDefaultMap?[product] = defaultsInstance
            }
        })
        return defaultsInstance as! Self
    }
    
    
    
    class func root(_ updateBlock: @escaping (_ root: [AnyHashable : Any]) -> Void) {
        let ud = UserDefaults.standard
        var rootImmutable = ud.dictionary(forKey: BO_SDK_ROOT_USER_DEFAULTS_KEY)
        if rootImmutable == nil {
            rootImmutable = [:]
        }
        let rootMutable = rootImmutable
        updateBlock(rootMutable ?? [:])
        ud.set(rootMutable, forKey: BO_SDK_ROOT_USER_DEFAULTS_KEY)
        ud.synchronize()
        
    }
    
    func updateDefault(forProduct updateBlock: @escaping (_ produtContainer: [AnyHashable : Any]) -> Bool) {
        weak var weakSelf = self
        BOFUserDefaults.root({ root in
            var rootObj = root
            //Get the product level defaults
            //Eg:
            // com.blotout.root.sdk
            //      -- Product1
            //          --Product1 defaults
            //      -- Product2
            //          --Product2 defaults
            
            var productContainer: [AnyHashable : Any]? = nil
            if let productKey = weakSelf?.productKey {
                productContainer = root[productKey] as? [AnyHashable : Any]
            }
            if productContainer == nil {
                productContainer = [:]
            }
            let hasChanged = updateBlock(productContainer ?? [:])
            if hasChanged {
                rootObj[weakSelf?.productKey] = productContainer
            }
        })
        
    }
    
    
    
    func setObject(_ obj: Any?, forKey aKey: String) {
        
        if obj == nil {
            return
            //We have batch updates in progress
        }
        if var productContainer = productContainer {
            productContainer[aKey as! AnyHashable] = obj
        } else {
            updateDefault(forProduct: { produtContainer in
                var productContainer = produtContainer
                productContainer[aKey as! AnyHashable] = obj
                return true
            })
        }
    }
    
    
    func removeObjectForKey( aKey: Any) {
        
        if self.productContainer != nil
        {
            self.productContainer?.removeValue(forKey: aKey as! AnyHashable)
        }
        else {
            
            updateDefault { produtContainer in
                var newContainer = produtContainer
                newContainer.removeValue(forKey: aKey as! AnyHashable)
                return true
            }
        }
    }
    func object(forKey key: String) -> Any? {
        var object: Any? = nil
        
        if let productContainer = productContainer {
            object = productContainer[key]
        } else {
            updateDefault(forProduct: { produtContainer in
                object = produtContainer[key]
                return false
            })
        }
        return object
    }
}

